### PR TITLE
Add bump-mapping support (feature #5304)

### PR DESCRIPTION
**Pull request by Capostrophic**
[Feature request](https://gitlab.com/OpenMW/openmw/issues/5304)

tl;dr: SHINY ROCKS BEGONE.

Add bump-mapping support using MGE XE fixed function pipeline emulation as a reference.
Bump-mapped objects always use shaders.

Morrowind Code Patch option where the lighting is applied to environment map reflections is added as well. Affected objects are rendered with shaders.

Notes:
1. Theoretically bump map references can be added to OSG format files. Bad things may happen since there's no NIF-specific information there. They shouldn't be used in them anyway since normal maps are better. I didn't update the relevant documentation part about texture types supported in OSG files due to this.
2. Spherical environment maps look somewhat different to Morrowind (I'm not sure how), and various texture generation types Morrowind and our fixed function pipeline has aren't replicated in shaders.
3. Normal maps can now only be picked up by autodetection, but some of the relevant code hasn't been removed as more recent NIF files include normal map information.
4. (Auto-detected) normal maps and bump maps can be used simultaneously as long as it's not literally the same file, as they're fundamentally different from the rendering viewpoint. If there's a conflict, only the bump map will be used.

I've updated the documentation, removing and replacing a lot of inaccurate and/or outdated information. It's still helpful in case the user wants to convert bump maps that are actually normal maps to be handled as normal maps.

Example (old version of Lougian's Hlaalu retexture + On The Rocks):
0.45.0/master:
screenshot199
With the option toggled off (vanilla-like)
screenshot205
With the option toggled on (MCP-like)
screenshot202

It's difficult to show it on pictures.

### CODE REVIEW

File: apps/opencs/model/world/data.cpp [hunk 1]
```diff
@@ -81,6 +81,7 @@ CSMWorld::Data::Data (ToUTF8::FromType encoding, bool fsStrict, const Files::Pat
     Shader::ShaderManager::DefineMap shadowDefines = SceneUtil::ShadowManager::getShadowsDisabledDefines();
     defines["forcePPL"] = "0";
     defines["clamp"] = "1";
+    defines["preLightEnv"] = "0";
```
Comment (AnyOldName3) [inline apps/opencs/model/world/data.cpp]: This could probably be more descriptive.

**Comment by AnyOldName3** [comment]
This looks like it disables normal maps if there's no environment map.

**Comment by Capostrophic** [comment]
No. Normal maps will still be autodetected, but they won't be picked up from Morrowind NIFs (because they're not supposed to be, they're not referenced there).

**Comment by AnyOldName3** [comment]
Derp. I read an `#if` in the shader wrong.

**Comment by akortunov** [comment]
> Has no effect in fixed function pipeline.

Are you sure that MCP behaves in the same way?

**Comment by Capostrophic** [comment]
Well, obviously it does, but do you know of a good way to do it in OpenGL?

**Comment by Capostrophic** [comment]
Added a workaround like for bump maps.

**Comment by psi29a** [comment]
Is this something nice for 0.46 in your opinion?

**Comment by Capostrophic** [comment]
I mean, yes, I did make a promise to add it last year (and to add it specifically in 2020), I consider it a 1.0 blocker and it's an awesome addition to all the mod compatibility improvements in 0.46.0 that are already present.

However, akortunov wanted some Russian modders to test it out, that may take a bit of time.

**Comment by psi29a** [comment]
Awaiting @akortunov stamp of approval for testing.

**Comment by akortunov** [comment]
> Awaiting @akortunov stamp of approval for testing.

I did not get any complaints from mentioned modders, so I suppose that this feature works.
Can not tell if this feature works as in Morrowind, though.

**Comment by AnyOldName3** [comment]
Is there anything we *could* do to get a direct comparison? Is ESSImporter good enough to get the right camera position and lighting conditions? I imagine this is very sensitive to both of those.

**Comment by Capostrophic** [comment]
> Is there anything we could do to get a direct comparison? 

Probably not. As I said in the starting post, our spherical environment mapping looks different to Morrowind (at least Morrowind under Wine, but its rendering should be fairly accurate) so even if the position and the looking angle are the same comparing bump mapping implementations will be difficult.

Which is confusing. Wine and OSG in theory do exactly the same OpenGL calls so spherical environment maps look the same in both FFP-based implementations.

:man_shrugging:

Would be nice to get an apitrace, but specifically on Windows and not under Wine.

**Comment by akortunov** [comment]
Merge, then?
